tools/kccm-flavor-gen/main.go

```diff
@@ -0,0 +1,190 @@
+		GeneratorOptions: &kustomizetypes.GeneratorOptions{
+			DisableNameSuffixHash: true,
+		},
+		ConfigMapGenerator: []kustomizetypes.ConfigMapArgs{
+			{
+				GeneratorArgs: kustomizetypes.GeneratorArgs{
+					Namespace: namespace,
+					Name:      "cloud-config",
+					KvPairSources: kustomizetypes.KvPairSources{
+						FileSources: []string{"cloud-config"},
+					},
+				},
+			},
+		},
+	}
+
+	if err := yaml.NewEncoder(kustomizationFile).Encode(&kustomization); err != nil {
+		return fmt.Errorf("failed decoding kustomization yaml file: %v", err)
+	}
+
+	cloudConfig := cloudprovider.CloudConfig{
+		LoadBalancer: cloudprovider.LoadBalancerConfig{
+			Enabled:              true,
+			CreationPollInterval: 30,
+		},
+		Namespace: namespace,
+	}
+
+	cloudConfigFile, err := os.Create(filepath.Join(kustomizationPath, "cloud-config"))
+	if err != nil {
+		return fmt.Errorf("failed creating kccm cloud config file: %v", err)
+	}
+
+	if err := yaml.NewEncoder(cloudConfigFile).Encode(&cloudConfig); err != nil {
+		return fmt.Errorf("failed encoding kccm cloud config to yaml: %v", err)
+	}
+
+	return nil
+}
+
+func main() {
+	var base = flag.String("base", "https://github.com/qinqon/cloud-provider-kubevirt/config/base?ref=kustomize-overlays-cm-secrets", "kccm kustomize base URL")
+	var templatePath = flag.String("template", "../../templates/cluster-template.yaml", "Cluster template to generate flavor for it will append 'kccm' to the file base name")
+	flag.Parse()
+
+	kustomizationPath, err := ioutil.TempDir("/tmp", "kustomize-overlay-capk-kccm")
+	if err != nil {
+		panic(err)
+	}
+	defer os.RemoveAll(kustomizationPath)
+
+	if err := kustomize(kustomizationPath, *base, "${CLUSTER_NAME}", "${NAMESPACE}"); err != nil {
+		panic(err)
+	}
+
+	fSys := filesys.MakeFsOnDisk()
+
+	k := krusty.MakeKustomizer(krusty.MakeDefaultOptions())
+	m, err := k.Run(fSys, kustomizationPath)
+	if err != nil {
+		panic(fmt.Errorf("failed running cloud provider kustomize: %v", err))
+	}
+	yamlTemplate, err := m.AsYaml()
+	if err != nil {
+		panic(fmt.Errorf("failed marshaling template as YAML: %v", err))
+	}
+	flavorPath := fmt.Sprintf("%s-kccm.yaml", strings.TrimSuffix(*templatePath, filepath.Ext(*templatePath)))
+	flavorPathFile, err := os.Create(flavorPath)
+	if err != nil {
+		panic(fmt.Errorf("failed creating kccm flavor file: %v", err))
+	}
+	clusterTemplate, err := ioutil.ReadFile(*templatePath)
+	if err != nil {
+		panic(fmt.Errorf("failed reading cluster template file: %v", err))
+	}
+
+	flavorPathFile.Write(clusterTemplate)
+	flavorPathFile.Write([]byte("---\n"))
+	flavorPathFile.Write(yamlTemplate)
+
+	// Add authorization to extension-apiserver-authorization-reading using
+	// a ClusterRoleBinding since clusterctl it's going to replace namespace
+	// "kube-system" at RoleBinding [1]
+	// [1] https://github.com/kubernetes-sigs/cluster-api/blob/6b042456383aa18ba61fc5de5c41901dd7267c60/cmd/clusterctl/client/repository/template.go#L131
+	flavorPathFile.Write([]byte(`---
```

Comment:
Why does cloud-provider-kubevirt need access to the kube-system namespace?

Reply:
the cloud-provider-kubevirt accesses the `extension-apiserver-authentication` configMap, which is in kube-system namespace.
The configmap contains client ca certificate.
We were unable to run kccm without giving it permissions to read that configmap, according to https://kubernetes.io/docs/tasks/extend-kubernetes/setup-extension-api-server/
Looks like kccm creates a client that uses the ca cert for authentication
https://github.com/kubernetes-sigs/apiserver-builder-alpha/blob/master/docs/concepts/auth.md#client-certificate-authentication

Reply:
> the cloud-provider-kubevirt accesses the extension-apiserver-authentication configMap, which is in kube-system namespace.

It doesn't make sense to me why we'd need to give access to this configmap in the infra cluster.

In which cluster (infra or guest) is this configmap access needed? My expectation is it would be the guest cluster, but i don't fully understand the purpose of needing the client CA.  I thought the kccm was just using reconcile loops and not acting as a server that needs to do any CA validation. maybe i'm misunderstanding something

Reply:
> In which cluster (infra or guest) is this configmap access needed?

The access is needed at infra cluster.

See the kccm logs, without and with provided access to the configmap:

```
I0817 11:31:20.084374       1 serving.go:348] Generated self-signed cert in-memory
I0817 11:31:20.374485       1 controllermanager.go:143] Version: v0.0.0-master+$Format:%H$
I0817 11:31:20.377956       1 secure_serving.go:210] Serving securely on [::]:10258
I0817 11:31:20.378223       1 leaderelection.go:248] attempting to acquire leader lease kube-system/cloud-controller-manager...
I0817 11:31:20.378548       1 requestheader_controller.go:169] Starting RequestHeaderAuthRequestController
I0817 11:31:20.378594       1 shared_informer.go:255] Waiting for caches to sync for RequestHeaderAuthRequestController
I0817 11:31:20.378600       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I0817 11:31:20.378651       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::client-ca-file"
I0817 11:31:20.378666       1 shared_informer.go:255] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0817 11:31:20.378672       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file"
I0817 11:31:20.378695       1 shared_informer.go:255] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0817 11:31:20.478845       1 shared_informer.go:262] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0817 11:31:20.478863       1 shared_informer.go:262] Caches are synced for RequestHeaderAuthRequestController
I0817 11:31:20.478843       1 shared_informer.go:262] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0817 11:31:37.836557       1 leaderelection.go:258] successfully acquired lease kube-system/cloud-controller-manager
I0817 11:31:37.836640       1 event.go:294] "Event occurred" object="kube-system/cloud-controller-manager" fieldPath="" kind="Lease" apiVersion="coordination.k8s.io/v1" type="Normal" reason="LeaderElection" message="kubevirt-cloud-controller-manager-66b64597c5-4rt2x_60531b2c-9cba-42a7-aea5-f357527cc085 became leader"
I0817 11:31:38.446279       1 controllermanager.go:291] Started "service"
W0817 11:31:38.446294       1 core.go:111] --configure-cloud-routes is set, but cloud provider does not support routes. Will not configure cloud provider routes.
W0817 11:31:38.446298       1 controllermanager.go:279] Skipping "route"
I0817 11:31:38.446385       1 controller.go:233] Starting service controller
I0817 11:31:38.446400       1 shared_informer.go:255] Waiting for caches to sync for service
I0817 11:31:38.446436       1 node_controller.go:118] Sending events to api server.
I0817 11:31:38.446489       1 controllermanager.go:291] Started "cloud-node"
I0817 11:31:38.446564       1 node_controller.go:157] Waiting for informer caches to sync
I0817 11:31:38.446608       1 node_lifecycle_controller.go:77] Sending events to api server
I0817 11:31:38.446639       1 controllermanager.go:291] Started "cloud-node-lifecycle"
I0817 11:31:38.546696       1 shared_informer.go:262] Caches are synced for service
E0817 11:31:38.555660       1 node_controller.go:249] Error getting instance metadata for node addresses: nodes "node01" is forbidden: User "system:serviceaccount:kvcluster:cloud-controller-manager" cannot get resource "nodes" in API group "" at the cluster scope
E0817 11:31:38.558487       1 node_controller.go:249] Error getting instance metadata for node addresses: nodes "node01" is forbidden: User "system:serviceaccount:kvcluster:cloud-controller-manager" cannot get resource "nodes" in API group "" at the cluster scope
......
```


And with access to the extension-apiserver-authentication configmap:
```
I0817 11:33:57.520132       1 serving.go:348] Generated self-signed cert in-memory
I0817 11:33:57.989834       1 controllermanager.go:143] Version: v0.0.0-master+$Format:%H$
I0817 11:33:57.992489       1 requestheader_controller.go:169] Starting RequestHeaderAuthRequestController
I0817 11:33:57.992517       1 shared_informer.go:255] Waiting for caches to sync for RequestHeaderAuthRequestController
I0817 11:33:57.992571       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::client-ca-file"
I0817 11:33:57.992587       1 shared_informer.go:255] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0817 11:33:57.992637       1 secure_serving.go:210] Serving securely on [::]:10258
I0817 11:33:57.992694       1 configmap_cafile_content.go:202] "Starting controller" name="client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file"
I0817 11:33:57.992705       1 shared_informer.go:255] Waiting for caches to sync for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0817 11:33:57.992874       1 leaderelection.go:248] attempting to acquire leader lease kube-system/cloud-controller-manager...
I0817 11:33:57.993010       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I0817 11:33:58.093118       1 shared_informer.go:262] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::requestheader-client-ca-file
I0817 11:33:58.093210       1 shared_informer.go:262] Caches are synced for RequestHeaderAuthRequestController
I0817 11:33:58.093218       1 shared_informer.go:262] Caches are synced for client-ca::kube-system::extension-apiserver-authentication::client-ca-file
I0817 11:34:13.789110       1 leaderelection.go:258] successfully acquired lease kube-system/cloud-controller-manager
I0817 11:34:13.789562       1 event.go:294] "Event occurred" object="kube-system/cloud-controller-manager" fieldPath="" kind="Lease" apiVersion="coordination.k8s.io/v1" type="Normal" reason="LeaderElection" message="kubevirt-cloud-controller-manager-66b64597c5-sjxzh_04e22a2e-e396-4539-bcc1-99d9335667cf became leader"
I0817 11:34:14.397761       1 node_lifecycle_controller.go:77] Sending events to api server
I0817 11:34:14.397957       1 controllermanager.go:291] Started "cloud-node-lifecycle"
I0817 11:34:14.399045       1 controllermanager.go:291] Started "service"
W0817 11:34:14.399077       1 core.go:111] --configure-cloud-routes is set, but cloud provider does not support routes. Will not configure cloud provider routes.
W0817 11:34:14.399094       1 controllermanager.go:279] Skipping "route"
I0817 11:34:14.399216       1 controller.go:233] Starting service controller
I0817 11:34:14.399239       1 shared_informer.go:255] Waiting for caches to sync for service
I0817 11:34:14.400402       1 node_controller.go:118] Sending events to api server.
I0817 11:34:14.400562       1 controllermanager.go:291] Started "cloud-node"
I0817 11:34:14.400787       1 node_controller.go:157] Waiting for informer caches to sync
I0817 11:34:14.501197       1 shared_informer.go:262] Caches are synced for service

```

Reply:
> E0817 11:31:38.555660       1 node_controller.go:249] Error getting instance metadata for node addresses: nodes "node01" is forbidden: User "system:serviceaccount:kvcluster:cloud-controller-manager" cannot get resource "nodes" in API group "" at the cluster scope
E0817 11:31:38.558487       1 node_controller.go:249] Error getting instance metadata for node addresses: nodes "node01" is forbidden: User "system:serviceaccount:kvcluster:cloud-controller-manager" cannot get resource "nodes" in API group "" at the cluster scope

That error is related to trying to access nodes, not a config map.  Also, we shouldn't need rbac for the infra nodes either. 

The only rbac necessary for the infra cluster should be a rolebinding for VirtualMachineInstances and Services within the cluster's namespace.

Reply:
This this is a dependency comming from the cloud-provider api, other cloud providers has this RBACs too https://raw.githubusercontent.com/kubernetes/cloud-provider-vsphere/master/manifests/controller-manager/cloud-controller-manager-role-bindings.yaml

Reply:
Disabling in cluster client authenticator config fix the issue, although I am not sure if it can create new ones https://github.com/kubevirt/cloud-provider-kubevirt/pull/141

Reply:
> This this is a dependency comming from the cloud-provider api, other cloud providers has this RBACs too

yes, but isn't this rbac supposed to be applied to the tenant cluster, and not the infra cluster? In this PR we're creating a clusterRole in the **infra** cluster to read the infra cluster's configman. I'd expect us to be needing access  to the tenant cluster's configmap, but maybe I'm mistaken?

Reply:
> > This this is a dependency comming from the cloud-provider api, other cloud providers has this RBACs too
> 
> yes, but isn't this rbac supposed to be applied to the tenant cluster, and not the infra cluster? In this PR we're creating a clusterRole in the **infra** cluster to read the infra cluster's configman. I'd expect us to be needing access to the tenant cluster's configmap, but maybe I'm mistaken?

We run the kccm on the infra so the cloud-provider is trying to access the one at infra, but looks like we don't need apiserver client authentication for the load balancer (the only feature kccm implement from cloud-provider).